info-compiler/src/test/java/com/google/sps/NewsContentExtractorTest.java

```diff
@@ -0,0 +1,133 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.webcrawler;
+
+import com.google.sps.data.NewsArticle;
+import java.io.InputStream;
+import java.io.IOException;
+import java.net.URL;
+import java.util.Optional;
+import org.apache.tika.exception.TikaException;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.xml.sax.SAXException;
+import static org.mockito.Mockito.*;
+
+/**
+ * A tester for news article content extraction.
+ */
+@RunWith(JUnit4.class)
+public final class NewsContentExtractorTest {
+  private final static String URL =
+    "https://www.cnn.com/2020/06/23/politics/aoc-ny-primary-14th-district/index.html";
```

Comment:
Thanks this makes sense! I will try

Reply:
I created a new test `extractContentFromHtml_alwaysValidWebpage()` that has a mock `InputStream` returning predefined content.
The content extractor repeatedly calls [`read(byte[], int, int)`](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#read-byte:A-int-int-) on the `InputStream`. So, I mocked the object to return -1 for `read()`, which is the indicator of the end of the stream. Then the extractor will parse and extract the empty content.
Do you think I should write sample HTML for the extractor and convert the HTML to `InputStream`? It's definitely doable, thought it will take some tweaking to define the "expected" content (since the implementation of the extractor is hidden from me).

Reply:
Yeah actually I would just mock the HtmlParser so that whatever it returns is fully controllable by you in these tests.

Reply:
Oh gotcha! That works too. Will do

Reply:
I'll make the class non-static first like `RelevancyChecker`

Reply:
I wrote `extractContentFromHtml_emptyDocParser()` that has a mock parser. The tricky part is that the parser doesn't directly return any results. The code looks like this;
```
      parser.parse(htmlFileStream, boilerpipeHandler, metadata);
      TextDocument textDocument = boilerpipeHandler.getTextDocument();
```
`boilerpipeHandler`, which is the input parameter, gets modified in `parser.parse()` and then returns `getTextDocument()`. By a invocation lambda function, I can get the reference to `boilerpipeHandler`. However the handler class doesn't seem to have a setter method that allows me to set its content, and I'm also not allowed to replace the original reference to point to a mock handler object.
Because of this situation, this method appears to be the best fit: 
[`endDocument()`](https://docs.oracle.com/javase/7/docs/api/org/xml/sax/ContentHandler.html#endDocument()). I called this method on `boilerpipeHandler` to indicate that the document has ended. As a result, the extracted content would predictably be empty. The constructed `NewsArticle` would be the same as `NEWS_ARTICLE_WITH_EMPTY_CONTENT`.

Reply:
Let me know if this is okay, or if there is a better way

Reply:
Looks like there's an "addTextBlock" inherited method? https://tika.apache.org/1.0/api/org/apache/tika/parser/html/BoilerpipeContentHandler.html

Reply:
and "setTitle"

Reply:
oh sad apparently "addTextBlock" is protected :/ I guess probably the only other way I can think to get around this is to overload extractContentFromHtml and have another version that also takes in a BoilerpipeContentHandler so you can mock it to return whatever TextDocument you want?

Reply:
sorry this got so complicated 😬

Reply:
No this was a good dive into Mockito!
Then I'll write another version of the method